src/common/thread/thread.cc

```diff
@@ -182,4 +171,113 @@ class ThreadPool {
     inline static std::mutex global_thread_pool_mutex_;
     constexpr static size_t kTaskQueueFactor = 16;
 };
+
+ThreadPool::ScopedOmpSetter::ScopedOmpSetter(int num_threads) {
+    if (global_build_thread_pool_size_ == 0) {  // this should not happen in prod
+        omp_before = omp_get_max_threads();
+    } else {
+        omp_before = global_build_thread_pool_size_;
+    }
+
+    omp_set_num_threads(num_threads <= 0 ? omp_before : num_threads);
+}
+
+ThreadPool::ScopedOmpSetter::~ScopedOmpSetter() {
+    omp_set_num_threads(omp_before);
+}
+
+void
+ExecOverSearchThreadPool(std::vector<std::function<void()>>& tasks) {
+    auto pool = ThreadPool::GetGlobalSearchThreadPool();
+    std::vector<folly::Future<folly::Unit>> futures;
+    futures.reserve(tasks.size());
+    for (auto&& t : tasks) {
+        futures.emplace_back(pool->push([&t]() {
+            ThreadPool::ScopedOmpSetter setter(1);
+            t();
+        }));
+    }
+    std::this_thread::yield();
+    // check for exceptions. value() is {}, so either
+    //   a call does nothing, or it throws an inner exception.
+    for (auto& f : futures) {
+        f.wait();
+        f.result().value();
```

Comment:
I'd rather wait for all futures to finish and then the theirs results tbh

Reply:
In our use case, the difference may not be significant, but upon closer consideration, checking after waiting does seem better.